### DISCUSSION TITLE
how to get the markdown snippet in macro function definition?

### DISCUSSION THREAD

**Discussion by jcyrss** [fralau/mkdocs-macros-plugin#210] [Q&A]
Some markdown contents only for some users. I know it could be done like this in markdown file,

```py
{% if site_type in ['usertype1', 'usertype2'] %}
  this is the content only for usertype #1 and #2
{% endif %}
```

But if I want to define a macro do that, like this,

```py
{{ only_for('usertype1', 'usertype2') }}
  this is the content only for usertype #1 and #2
{{ end only_for}}
```

How to define end of the `onlyfor` scope in markdown file? like `{{ end only_for}}` ?

How to get the markdown snippet in macro function definition?

```py
def define_env(env):

    @env.macro
    def only_for():
        snippet = # how to get the markdown snippet?
        if env.variables.site_type in ['usertype1', 'usertype2']:
            return snippet 
```

**Comment by fralau** [comment]
Thanks for your question. I think I understand what you wish to do. Alas, Jinja2 does not allow (at least to my knowledge) to know what's "in-between".

To solve that problem, I am not sure you that creating snippets would give you a simple solution, since you would have to store them somewhere, etc. Do you have a special motivation for that?

What I would do would be the following:

in extra in the config file (or in the page header):

```yaml
extra:
  authorized_users: usertype1, usertype2
```

```
{% if site_type in authorized_users %}
  this is the content only for usertype #1 and #2
{% endif %}
```

In that way, if you wanted to change the number of users, or their names, etc., you would only have to change your config file.

**Reply by jcyrss** [reply]
Thank you for your suggestion.

> Jinja2 does not allow (at least to my knowledge) to know what's "in-between".

What a pity. Sometimes, we need more complex handle logic, which depends on the content of "in-between". 

Like, for the following html snippet, I want to automatically append a frame with that html snippet embeded to show the visual result. 

```html
{{< htmlpg >}}
<head>
  <style>
  .wrapper{
    display:flex; 
    gap: .5em;
  }
  .wrapper > div {
    border: 1px solid teal;
  }
  </style>
</head>  
<body>
  <div class="wrapper">  
    <div>1</div>
    <div>2</div>
    <div> 3
      <br>3
    </div>
    <div>6</div>
    <div>7</div>
    <div>8888888888888</div>
    <div>99999999999999999999999999999999999</div>
  </div>
</body>  
{{< /htmlpg >}}
```

<br>

Hugo shortcode template could do that, it gets the in-between content just by `{{ .Inner }}`

That would be convient if mkdoc could do like that too. 

<br>

But I think it could be feasible to put that work in `on_pre_page_macros` to change the content of the markdown.

```py
def on_pre_page_macros(env):
    # search through the env.markdown and modify it as we need
    env.markdown = handle_htmlpg(env.markdown)
```

But would that slow down the speed of mkdoc processing?

**Reply by fralau** [reply]
> What a pity. Sometimes, we need more complex handle logic, which depends on the content of "in-between".

Agreed.

> Hugo shortcode template could do that, it gets the in-between content just by {{ .Inner }}

I am interested in that concept. Would have link or some material, so that I can understand how it works?

**Reply by jcyrss** [reply]
> Hugo shortcode template could do that, it gets the in-between content just by {{ .Inner }}
> ===========
> I am interested in that concept. Would have link or some material, so that I can understand how it works?

Check it out

https://gohugo.io/methods/shortcode/inner/

**Comment by fralau** [comment]
@jcyrss Thanks a lot for this information. I looked up what a [shortcode](https://gohugo.io/content-management/shortcodes/#use-shortcodes) is in Hugo, and it seems that it is functionally equivalent to an MkDocs macro, right?

I had toyed in the past with the concept of "shortcodes with opening and closing tag". If one wants to implement that elegantly, one obviously needs a stack that keeps track of the tags (because they can be embedded in each other). Basically you have sets of twin macros (opening and end), which rely on some scaffolding. It is not difficult to implement that (I did it before), it just requires some precision work.

**But accessing the "inner"**, as Hugo is doing so elegantly... I had thought about it, and it might definitely have some applications.

Right now, I am not sure how we could do that. We might have, as you suggest, to do some parsing of the markdown.

And furthermore, if we managed to do it as a [pluglet](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pluglets/), rather than a core function of Mkdocs-Macros...

Basically we should start by writing a module 🤔 . It wouldn't be any old module, though, but something more sophisticated.  As you see, I went full circle and back to your initial idea. 🙂

**Comment by fralau** [comment]
I asked [Ma](https://www.bookseriesinorder.com/big-sigma/) (ChatGPT) the following question:

> Is there a way to know, with Jinja2, what code (HTML, markdown, etc.) is between two Jinja2 directives?

**Answer**

In Jinja2, the content between two directives is processed when the template is rendered. However, Jinja2 does not provide a built-in way to directly access the raw, unprocessed code between two directives.

For example, if you have a block in your Jinja2 template like this:

```
{% block content %}
    <h1>Hello, World!</h1>
{% endblock %}
```
The string `<h1>Hello, World!</h1>` is the code between the `{% block content %}` and `{% endblock %}` directives. But Jinja2 does not provide a built-in function to extract this string directly.

If you need to manipulate or analyze the raw template code, you might need to read the template file as a regular text file in Python, and then use standard string manipulation or regular expressions to extract the content between the directives.

Please note that this approach bypasses Jinja2’s template processing and works directly on the raw template code. It might not be suitable for all use cases, especially if your templates use advanced Jinja2 features or inherit from other templates.

____

My comment: Pretty to the point, I must say.

**Comment by fralau** [comment]
> But I think it could be feasible to put that work in `on_pre_page_macros` to change the content of the markdown.
> 
> ```python
> def on_pre_page_macros(env):
>     # search through the env.markdown and modify it as we need
>     env.markdown = handle_htmlpg(env.markdown)
> ```
> 
> But would that slow down the speed of MkDocs processing?

Yes, it would, a little of course, though I wouldn't expect that to be a huge problem.

I guess it would be possible to build a data structure (tree) containing the open blocks and their contents (or how to access it, in case Jinja2 allows it).

**Comment by fralau** [comment]
@jcyrss By the way (lateral thinking) have you had a look at [Template Inheritance](https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance), in Jinja2?

It might solve your problem, as well? I can't tell, because I don't have the full statement of the problem you want to solve, and your constraints.
